fs/parallel_unpacker.go

```diff
@@ -0,0 +1,143 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package fs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/containerd/containerd/archive"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/errdefs"
+	"github.com/containerd/log"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"golang.org/x/sync/errgroup"
+)
+
+const defaultRetryPeriod = 500 * time.Millisecond
+
+type parallelLayerUnpacker struct {
+	*layerUnpacker
+	controller            LayerUnpackResourceController
+	discardUnpackedLayers bool
+}
+
+func NewParallelLayerUnpacker(fetcher Fetcher, archive Archive, controller LayerUnpackResourceController, discardUnpackedLayers bool) *parallelLayerUnpacker {
+	return &parallelLayerUnpacker{
+		layerUnpacker: &layerUnpacker{
+			fetcher: fetcher,
+			archive: archive,
+		},
+		controller:            controller,
+		discardUnpackedLayers: discardUnpackedLayers,
+	}
+}
+
+func (lu *parallelLayerUnpacker) Unpack(ctx context.Context, desc ocispec.Descriptor, mountpoint string, mounts []mount.Mount) error {
+	rc, local, err := lu.fetcher.Fetch(ctx, desc)
+	if err != nil {
+		return fmt.Errorf("cannot fetch layer: %w", err)
+	}
+	defer rc.Close()
+
+	release, err := lu.controller.AcquireUnpackLease(ctx)
+	if err != nil {
+		return err
+	}
+	defer release()
+
+	var errGroup errgroup.Group
+	if !local && !lu.discardUnpackedLayers {
+		var fetchLocalContentStoreReader bool
+
+		ingestReader, err := lu.controller.GetUnpackIngestReader()
+		if err != nil && errors.Is(err, ErrLayerIngestDoesNotExist) {
+			// Compressed layer is not on disk; so use the remote reader to download and store it in the content store.
```

Comment:
Do you really see this happening before? I think it's ok to fail here instead of performing another layer fetch.